packages/trackerless-network/src/logic/RandomGraphNode.ts

```diff
@@ -0,0 +1,297 @@
+            new RemoteRandomGraphNode(
+                descriptor,
+                this.config.randomGraphId,
+                toProtoRpcClient(new NetworkRpcClient(this.config.rpcCommunicator.getRpcClientTransport()))
+            )
+        ))
+        if (this.config.targetNeighbors.size() < this.config.numOfTargetNeighbors) {
+            this.config.neighborFinder.start()
+        }
+    }
+
+    private removedRandomContact(_removedDescriptor: PeerDescriptor, randomPeers: PeerDescriptor[]): void {
+        logger.trace(`New nearby contact found`)
+        if (this.stopped) {
+            return
+        }
+        this.config.randomContactPool!.replaceAll(randomPeers.map((descriptor) =>
+            new RemoteRandomGraphNode(
+                descriptor,
+                this.config.randomGraphId,
+                toProtoRpcClient(new NetworkRpcClient(this.config.rpcCommunicator.getRpcClientTransport()))
+            )
+        ))
+    }
+
+    private onPeerDisconnected(peerDescriptor: PeerDescriptor): void {
+        if (this.config.targetNeighbors.hasPeer(peerDescriptor)) {
+            this.config.targetNeighbors.remove(peerDescriptor)
+            this.config.connectionLocker.unlockConnection(peerDescriptor, this.config.randomGraphId)
+            this.config.neighborFinder.start([keyFromPeerDescriptor(peerDescriptor)])
+        }
+    }
+
+    private getNewNeighborCandidates(): PeerDescriptor[] {
+        return this.config.layer1.getNeighborList().getClosestContacts(this.config.peerViewSize).map((contact: DhtPeer) => {
+            return contact.getPeerDescriptor()
+        })
+    }
+
+    public hasProxyConnection(peerKey: PeerIDKey): boolean {
+        if (this.config.proxyConnectionServer) {
+            return this.config.proxyConnectionServer.hasConnection(peerKey)
+        }
+        return false
+    }
+
+    stop(): void {
+        if (!this.started) {
+            return
+        }
+        this.stopped = true
+
+        this.abortController.abort()
+        this.config.proxyConnectionServer?.stop()
+        this.config.targetNeighbors.getPeers().map((remote) => remote.leaveStreamNotice(this.config.ownPeerDescriptor))
+        this.config.rpcCommunicator.stop()
+        this.removeAllListeners()
+        this.config.layer1.off('newContact', (peerDescriptor, closestTen) => this.newContact(peerDescriptor, closestTen))
+        this.config.layer1.off('contactRemoved', (peerDescriptor, closestTen) => this.removedContact(peerDescriptor, closestTen))
+        this.config.layer1.off('newRandomContact', (peerDescriptor, randomPeers) => this.newRandomContact(peerDescriptor, randomPeers))
+        this.config.layer1.off('randomContactRemoved', (peerDescriptor, randomPeers) => this.removedRandomContact(peerDescriptor, randomPeers))
+        this.config.P2PTransport.off('disconnected', (peerDescriptor: PeerDescriptor) => this.onPeerDisconnected(peerDescriptor))
+        this.config.nearbyContactPool.stop()
```

Comment:
Would it be convenient if each of these dependencies would know the abortController of this node, and stop itself when the node aborts? It would reduce the coupling and dependency management between these components.